### PR TITLE
Migrate to BLS signature for consensus and other crypto signatures

### PR DESCRIPTION
**Pull request by rlan35**
Tested locally:

2 shards, 112 consensus, 316 total TPS, 11 nodes 127.0.0.1, 56, 157.478

### CODE REVIEW

File: scripts/go_executable_build.sh [hunk 1]
```diff
@@ -12,6 +12,7 @@ GOOS=linux
 GOARCH=amd64
 FOLDER=/${WHOAMI:-$USER}
 RACE=
+source ~/.bash_profile
```
Comment (LeoHChen) [inline scripts/go_executable_build.sh]: this won't work. please just export the variables here.

File: scripts/travis_checker.sh [hunk 2]
```diff
@@ -33,8 +33,10 @@ dirnames() {
 		-e 's:^\./::'
 }
 
+
 go_dirs="${tmpdir}/go_dirs.txt"
 dirnames < "${go_files}" | sort -u -t/ > "${go_dirs}"
+source ~/.bash_profile
```
Comment (LeoHChen) [inline scripts/travis_checker.sh]: this won't work, please just export your variables here.

File: test/deploy.sh [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 ROOT=$(dirname $0)/..
 USER=$(whoami)
+source ~/.bash_profile
```
Comment (LeoHChen) [inline test/deploy.sh]: please export your variables here.

File: test/crypto/bls/main.go [hunk 4]
```diff
@@ -13,9 +17,17 @@ func main() {
...
-		bls.Init(bls.BLS12_381)
 		var sec bls.SecretKey
 		sec.SetByCSPRNG()
+
+		if i == 0 {
+			testECKey, _ := ecdsa.GenerateKey(crypto.S256(), rand.Reader)
+			log.Printf("Secret Key: 0x%s", sec.GetHexString())
+			log.Printf("Secret Key: 0x%s", hex.EncodeToString(sec.GetLittleEndian()))
+			log.Printf("Secret Key Length: %d", len(sec.GetLittleEndian()))
+			log.Printf("Secret Key: 0x%s", hex.EncodeToString(testECKey.D.Bytes()))
```
Comment (mikedoan) [inline test/crypto/bls/main.go]: what diff between testECKey and sec? one generated by ecdsa and one generated by bls

File: scripts/travis_checker.sh (hunk 2 above)
Comment (rlan35) [reply scripts/travis_checker.sh]: Done.

File: test/deploy.sh (hunk 3 above)
Comment (rlan35) [reply test/deploy.sh]: Done.

File: test/crypto/bls/main.go (hunk 4 above)
Comment (rlan35) [reply test/crypto/bls/main.go]: This is just testing code. I am just checking whether they are different. Turn out no difference.

**Comment by alajko** [comment]
is `bls/ffi/go` commited somewhere? I don't see it in master. Am I missing something?

**Comment by rlan35** [comment]
> is `bls/ffi/go` commited somewhere? I don't see it in master. Am I missing something?

It's in the forked repo harmony-one/bls

**Comment by LeoHChen** [comment]
I've created a patch to fix the build issue on Jenkins server.  I tested it works.

```
diff --git a/scripts/go_executable_build.sh b/scripts/go_executable_build.sh
index f3ddde7..380a981 100755
--- a/scripts/go_executable_build.sh
+++ b/scripts/go_executable_build.sh
@@ -13,9 +13,10 @@ GOARCH=amd64
 FOLDER=/${WHOAMI:-$USER}
 RACE=

-export CGO_CFLAGS="-I$PWD/../bls/include -I$PWD/../mcl/include"
-export CGO_LDFLAGS="-L$PWD/../bls/lib"
-export LD_LIBRARY_PATH=$PWD/../bls/lib:$PWD/../mcl/lib
+HMY_PATH=$GOPATH/src/github.com/harmony-one
+export CGO_CFLAGS="-I$HMY_PATH/bls/include -I$HMY_PATH/mcl/include"
+export CGO_LDFLAGS="-L$HMY_PATH/bls/lib"
+export LD_LIBRARY_PATH=$HMY_PATH/bls/lib:$HMY_PATH/mcl/lib

 OS=$(uname -s)
 case $OS in
@@ -27,6 +28,10 @@ case $OS in
       ;;
 esac

+echo $CGO_CFLAGS
+echo $CGO_LDFLAGS
+echo $LD_LIBRARY_PATH
+
 if [ "$(uname -s)" == "Darwin" ]; then
    MD5='md5 -r'
    GOOS=darwin
```

File: consensus/consensus_leader.go [hunk 5]
```diff
@@ -143,12 +120,15 @@ func (consensus *Consensus) processCommitMessage(message consensus_proto.Message
...
-		utils.GetLogInstance().Warn("Failed to marshal the announce message", "error", err)
-	}
-	if schnorr.Verify(crypto.Ed25519Curve, value.PubKey, messageBytes, signature) != nil {
-		utils.GetLogInstance().Warn("Received message with invalid signature", "validatorKey", consensus.leader.PubKey, "consensus", consensus)
-		return
+		utils.GetLogInstance().Warn("Failed to marshal the prepare message", "error", err)
 	}
+	_ = messageBytes
+	_ = signature
+	// TODO: verify message signature
```
Comment (LeoHChen) [inline consensus/consensus_leader.go]: please open an ISSUE on this TODO

File: consensus/consensus_leader.go [hunk 6]
```diff
@@ -163,91 +143,57 @@ func (consensus *Consensus) processCommitMessage(message consensus_proto.Message
...
-	}
+	prepareSigs := consensus.prepareSigs
+	prepareBitmap := consensus.prepareBitmap
 
 	// proceed only when the message is not received before
-	_, ok = (*commitments)[validatorID]
+	_, ok = (*prepareSigs)[validatorID]
 	shouldProcess := !ok
-	if len((*commitments)) >= ((len(consensus.PublicKeys)*2)/3 + 1) {
+	if len((*prepareSigs)) >= ((len(consensus.PublicKeys)*2)/3 + 1) {
```
Comment (LeoHChen) [inline consensus/consensus_leader.go]: better to refactor this check into a function call, maybe open an issue.

File: consensus/consensus_leader.go [hunk 7]
```diff
@@ -163,91 +143,57 @@ func (consensus *Consensus) processCommitMessage(message consensus_proto.Message
...
-		point := crypto.Ed25519Curve.Point()
-		point.UnmarshalBinary(commitment)
-		(*commitments)[validatorID] = point
-		utils.GetLogInstance().Debug("Received new commit message", "num", len(*commitments), "validatorID", validatorID, "PublicKeys", len(consensus.PublicKeys))
+		var sign bls.Sign
+		err := sign.Deserialize(prepareSig)
+		if err != nil {
+			utils.GetLogInstance().Error("Failed to deserialize bls signature", "validatorID", validatorID)
+		}
+		// TODO: check bls signature
```
Comment (LeoHChen) [inline consensus/consensus_leader.go]: issue please.

File: consensus/consensus_leader.go [hunk 8]
```diff
@@ -279,142 +225,95 @@ func (consensus *Consensus) processResponseMessage(message consensus_proto.Messa
...
-		utils.GetLogInstance().Warn("Failed to marshal the announce message", "error", err)
-	}
-	if schnorr.Verify(crypto.Ed25519Curve, value.PubKey, messageBytes, signature) != nil {
-		utils.GetLogInstance().Warn("Received message with invalid signature", "validatorKey", consensus.leader.PubKey, "consensus", consensus)
-		return
+		utils.GetLogInstance().Warn("Failed to marshal the commit message", "error", err)
 	}
+	_ = messageBytes
+	_ = signature
+	// TODO: verify message signature
```
Comment (LeoHChen) [inline consensus/consensus_leader.go]: issue please.

File: consensus/consensus_validator.go [hunk 9]
```diff
@@ -62,10 +57,13 @@ func (consensus *Consensus) processAnnounceMessage(message consensus_proto.Messa
 	if err != nil {
 		utils.GetLogInstance().Warn("Failed to marshal the announce message", "error", err)
 	}
-	if schnorr.Verify(crypto.Ed25519Curve, consensus.leader.PubKey, messageBytes, signature) != nil {
-		utils.GetLogInstance().Warn("Received message with invalid signature", "leaderKey", consensus.leader.PubKey, "consensus", consensus)
-		return
-	}
+	_ = signature
+	_ = messageBytes
+	// TODO: verify message signature
```
Comment (LeoHChen) [inline consensus/consensus_validator.go]: issue

File: consensus/consensus_validator.go [hunk 10]
```diff
@@ -319,12 +246,54 @@ func (consensus *Consensus) processCollectiveSigMessage(message consensus_proto.
...
+	// Construct and send the prepare message
+	msgToSend := consensus.constructPrepareMessage()
 
 	consensus.SendMessage(consensus.leader, msgToSend)
 
-	// Set state to CommitDone
-	consensus.state = FinalCommitDone
+	consensus.state = PrepareDone
+
+	// TODO: the block catch up logic is a temporary workaround for full failure node catchup. Implement the full node catchup logic
```
Comment (LeoHChen) [inline consensus/consensus_validator.go]: please open an issue.

File: consensus/consensus_validator.go [hunk 11]
```diff
@@ -319,12 +246,54 @@ func (consensus *Consensus) processCollectiveSigMessage(message consensus_proto.
...
+	// The logic is to roll up to the latest blocks one by one to try catching up with the leader.
+	for {
+		val, ok := consensus.blocksReceived[consensus.consensusID]
+		if ok {
+			delete(consensus.blocksReceived, consensus.consensusID)
+
+			consensus.blockHash = [32]byte{}
```
Comment (LeoHChen) [inline consensus/consensus_validator.go]: use blockHash,

File: crypto/bls/bls.go [hunk 12]
```diff
@@ -0,0 +1,225 @@
+package bls
```
Comment (LeoHChen) [inline crypto/bls/bls.go]: there is no unit test of bls.go, please add it.

File: internal/utils/utils.go [hunk 13]
```diff
@@ -73,6 +74,21 @@ func GenKey(ip, port string) (kyber.Scalar, kyber.Point) {
 	return priKey, pubKey
 }
 
+// GenKeyBLS generates a bls key pair given ip and port.
+func GenKeyBLS(ip, port string) (*bls.SecretKey, *bls.PublicKey) {
```
Comment (LeoHChen) [inline internal/utils/utils.go]: no test program?

File: consensus/consensus_leader.go [hunk 14]
```diff
@@ -163,91 +143,57 @@ func (consensus *Consensus) processCommitMessage(message consensus_proto.Message
...
-		consensus.responseByLeader(challengeScalar, targetState == ChallengeDone)
+		// Construct prepared message
+		msgToSend, aggSig := consensus.constructPreparedMessage()
+		consensus.aggregatedPrepareSig = aggSig
 
-		// Broadcast challenge message
+		// Broadcast prepared message
 		host.BroadcastMessageFromLeader(consensus.host, consensus.GetValidatorPeers(), msgToSend, consensus.OfflinePeers)
 
 		// Set state to targetState (ChallengeDone or FinalChallengeDone)
```
Comment (chaosma) [inline consensus/consensus_leader.go]: targetState is prepare or commit now

File: consensus/consensus_leader.go [hunk 15]
```diff
@@ -279,142 +225,95 @@ func (consensus *Consensus) processResponseMessage(message consensus_proto.Messa
...
+		commitBitmap.SetKey(value.PubKey, true)
 	}
 
 	if !shouldProcess {
-		utils.GetLogInstance().Debug("Received new response message", "validatorID", strconv.Itoa(int(validatorID)))
+		utils.GetLogInstance().Debug("Received additional new commit message", "validatorID", strconv.Itoa(int(validatorID)))
 		return
 	}
 
 	threshold := 2
```
Comment (chaosma) [inline consensus/consensus_leader.go]: we don't need threshold = 1 anymore? If that's case, we can just remove this variable

File: consensus/consensus_validator.go [hunk 16]
```diff
@@ -174,106 +168,41 @@ func (consensus *Consensus) processChallengeMessage(message consensus_proto.Mess
...
-					continue
-				}
-
-			} else {
-				break
-			}
 
-		}
-	}
+	consensus.state = CommitDone
```
Comment (chaosma) [inline consensus/consensus_validator.go]: Why processPrepareMessage will set the state CommitDone? Not PreparedDone?

File: consensus/consensus_validator.go [hunk 17]
```diff
@@ -319,12 +246,54 @@ func (consensus *Consensus) processCollectiveSigMessage(message consensus_proto.
...
+	// TODO: verify multi signature
+
+	// Construct and send the prepare message
+	msgToSend := consensus.constructPrepareMessage()
 
 	consensus.SendMessage(consensus.leader, msgToSend)
 
-	// Set state to CommitDone
-	consensus.state = FinalCommitDone
+	consensus.state = PrepareDone
```
Comment (chaosma) [inline consensus/consensus_validator.go]: Shoule be PrepareDone or CommitDone?

**Comment by chaosma** [comment]
It looks good to me. Just a few comments.  BLS is much cleaner than Schnorr.

File: consensus/consensus_validator.go (hunk 16 above)
Comment (rlan35) [reply consensus/consensus_validator.go]: I will correct this in a following PR.

File: consensus/consensus_validator.go (hunk 17 above)
Comment (rlan35) [reply consensus/consensus_validator.go]: I will correct this in a following PR.

File: consensus/consensus_leader.go (hunk 15 above)
Comment (rlan35) [reply consensus/consensus_leader.go]: I will correct this in a following PR.

File: consensus/consensus_leader.go (hunk 14 above)
Comment (rlan35) [reply consensus/consensus_leader.go]: I will correct this in a following PR.

**Pull request merged**